benchmark/utils/general.hpp

```diff
@@ -434,4 +461,232 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * ```
+ * auto timer = get_timer(...);
+ * IterationControl ic(timer);
+ * for(auto status: ic.[warmup_run|run](manage_timings [default is true])){
+ *   if(! manage_timings) timer->tic();
+ *   // execute benchmark
+ *   if(! manage_timings) timer->toc();
+ * }
+ * ```
+ * At the beginning of both methods, the timer is reset.
+ * The `status` object exposes the member
+ * - `cur_it`, containing the current iteration number,
+ * and the methods
+ * - `is_finished`, checks if the benchmark is finished,
+ */
+class IterationControl {
+    using IndexType = unsigned int;  //!< to be compatible with GFLAGS type
+
+    class run_control;
+
+public:
+    /**
+     * Creates an `IterationControl` object.
+     *
+     * Uses the commandline flags to setup the stopping criteria for the
+     * warmup and timed run.
+     *
+     * @param timer the same timer that is to be used for the timings
+     */
+    explicit IterationControl(const std::shared_ptr<Timer> &timer)
+    {
+        status_warmup_ = {TimerManager{timer, false}, FLAGS_warmup,
+                          FLAGS_warmup, 0., 0};
+        if (FLAGS_repetitions == "auto") {
+            status_run_ = {TimerManager{timer, true}, FLAGS_min_repetitions,
+                           FLAGS_max_repetitions, FLAGS_min_runtime};
+        } else {
+            const auto reps =
+                static_cast<unsigned int>(std::stoi(FLAGS_repetitions));
+            status_run_ = {TimerManager{timer, true}, reps, reps, 0., 0};
+        }
+    }
+
+    IterationControl() = default;
+    IterationControl(const IterationControl &) = default;
+    IterationControl(IterationControl &&) = default;
+
+    /**
+     * Creates iterable `run_control` object for the warmup run.
+     *
+     * This run uses always a fixed number of iterations.
+     */
+    run_control warmup_run()
+    {
+        status_warmup_.cur_it = 0;
+        status_warmup_.timer->clear();
+        return run_control{&status_warmup_};
+    }
+
+    /**
+     * Creates iterable `run_control` object for the timed run.
+     *
+     * This run may be adaptive, depending on the commandline flags.
+     *
+     * @param manage_timings If true, the timer calls (`tic/toc`) are handled
+     * by the `run_control` object, otherwise they need to be executed outside
+     */
+    run_control run(bool manage_timings = true)
+    {
+        status_run_.cur_it = 0;
+        status_run_.timer->clear();
+        status_run_.timer->manage_timings = manage_timings;
+        return run_control{&status_run_};
+    }
+
+
+    double compute_average_time() const
+    {
+        return status_run_.timer->get_total_time() / get_num_repetitions();
+    }
+
+    IndexType get_num_repetitions() const { return status_run_.cur_it; }
+
+private:
+    struct TimerManager {
+        std::shared_ptr<Timer> timer;
+        bool manage_timings = false;
+
+        void tic()
+        {
+            if (manage_timings) timer->tic();
```

Comment:
```suggestion
            if (manage_timings) {
                timer->tic();
            }
```
also apply to next one from the gko [ref](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#control-flow-constructs)

Reply:
Is that part of the .clang-format specification? If not, perhaps it should be added there.

Reply:
it is not in current .clang-format. we only mention it in Contribution guideline.
Does clang-format support this after 6?

Reply:
After a bit of digging, it seems like clang-format (up to 13) still does not support this. But there is an PR for this here: https://reviews.llvm.org/D95168
So it seems like that some future version will support this.

Reply:
There is also a workaround using clang-tidy, but that is over kill (https://stackoverflow.com/a/28437960)